### PR TITLE
Hclang 104

### PR DESCRIPTION
**Pull request by Lambcreek**
### Description
Added more built in functions - should cover everything that we need to call from C++.

### Changelog
 - added "at" function
 - added "subList" function
 - added "+" function for both list and text types
 - added ftoa() C++ function which is called in "toText"

### Issues solved 
 - HCLANG-104

### Issues opened as a result of this 
None right now

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: code/cppRawFtoaFunction/ftoa.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
+
+#ifndef FTOA_H
+#define FTOA_H
+
+#include <cstdlib>
+#define MAX_STR_LEN 16
+
+char *ftoa(double d, int precision) {
```
Comment (C0DK) [inline code/cppRawFtoaFunction/ftoa.h]: needs header description

File: code/cppRawFtoaFunction/ftoa.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            d *= -1;
+            wholePart *= -1;
+        }
+
+        double fraction = d - wholePart;
+        int endOfBuffer = (int)buffer + MAX_STR_LEN;
+        for (; precision > 0 && (int)endOfString < endOfBuffer; precision--) {
+            fraction *= 10;
+            wholePart = (long)fraction;
+            *endOfString++ = '0' + (char)wholePart;
```
Comment (C0DK) [inline code/cppRawFtoaFunction/ftoa.h]: i get what u doin here, but maybe add a comment cuz i was confused at first thinking u wanted to concat.

File: code/src/main/kotlin/builtins/HclBuiltinFunctions.kt [hunk 3]
```diff
@@ -77,15 +80,51 @@ private inline fun<reified V, reified H, reified R> buildOperator(functionName:
...
+)
+
+private fun buildTextConcatFunction() = buildFunction(
+        identifier = "+",
+        parameters = listOf(
+                Parameter("leftHand", Type.Text),
+                Parameter("rightHand", Type.Text)
+        ),
+        returnType = Type.Text,
+        body = "char *ret = malloc((strlen(leftHand) + strlen(rightHand) + 1) * sizeof(char));\n" +
```
Comment (C0DK) [inline code/src/main/kotlin/builtins/HclBuiltinFunctions.kt]: memory leak :P

File: code/src/test/kotlin/builtinTests/TestAllBuiltinsExist.kt [hunk 4]
```diff
@@ -18,8 +18,8 @@ object TestAllBuiltinsExist : Spek({
                 "-",
                 "/",
                 "*",
-                "==",
-                "!=",
+                "equals",
+                "notEquals",
                 "<",
                 ">",
                 "&&",
```
Comment (Abildgaard96) [inline code/src/test/kotlin/builtinTests/TestAllBuiltinsExist.kt]: ville vi ændre dem her til 'and' & 'or' istedet for symbolerne?

Comment (Abildgaard96) [reply code/src/test/kotlin/builtinTests/TestAllBuiltinsExist.kt]: ps, syntes vi skulle overveje at ændre 'negated' tilbage til 'not' 
tror bare det er simplerer

**Pull request merged**